### PR TITLE
`many_buttons` enhancements

### PR DESCRIPTION
**Pull request by ickshonpe**
# Objective

`many_buttons` enhancements:
* use `argh` to manage the commandline arguments like the other stress tests
* add an option to set the number of buttons
* add a grid layout option
* centre the grid properly
* use viewport coords for the layout's style constraints
* replace use of absolute positioning

includes the changes from #9636

Displaying an image isn't actually about stress testing image rendering. Without a second texture (the first is used by the text) the entire grid will be drawn in a single batch. The extra texture used by the image forces the renderer to break up the batches at every button displaying an image, where it has to switch between the font atlas texture and the image texture.

## Solution

many_buttons_new

---

## Changelog

 `many_buttons` stress test example enhancements:
* uses `argh` to the manage the commandline arguments.
* New commandline args:
  - `--help` display info & list all commandline options
  - `--buttons`  set the number of buttons.
  - `--image-freq` set the frequency of buttons displaying images
  - `--grid` use a grid layout
* style constraints are specified in viewport coords insead of percentage values
* margins and nested bundles are used to construct the layout, instead of absolute positioning
* the button grid centered in the window, the empty gap along the bottom and right is removed
* an image is drawn as the background to every Nth button where N is set using the `--image-freq` commandline option.

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]
The CLI interface looks nice :) And the enhancements will be useful for more precise benchmarking without exploding our set of stress tests.

File: examples/stress_tests/many_buttons.rs [hunk 1]
```diff
@@ -64,50 +89,119 @@ fn button_system(
...
+    let as_rainbow = |i: usize| Color::hsl((i as f32 / buttons_f) * 360.0, 0.9, 0.8);
     commands.spawn(Camera2dBundle::default());
     commands
         .spawn(NodeBundle {
             style: Style {
+                display: Display::Grid,
                 width: Val::Percent(100.),
+                height: Val::Percent(100.0),
+                grid_template_columns: vec![RepeatedGridTrack::flex(args.buttons as u16, 1.0)],
+                grid_template_rows: vec![RepeatedGridTrack::flex(args.buttons as u16, 1.0)],
```
Comment (rparrett) [inline examples/stress_tests/many_buttons.rs:182]:
```suggestion
                grid_template_columns: RepeatedGridTrack::flex(args.buttons as u16, 1.0),
                grid_template_rows: RepeatedGridTrack::flex(args.buttons as u16, 1.0),
```

`Vec<RepeatedGridTrack>` has `From<RepeatedGridTrack>`

**Comment by ickshonpe** [comment]
> Some thoughts:
> 
> * Any particular reason to not be using the default window size here? The online [examples page](https://bevyengine.org/examples/#stress-tests) just looks nicer when everything is the same aspect ratio.

Not really, with a square window you get square buttons but it doesn't matter. I hadn't considered the examples page, I'll change it to the default.

**Comment by ickshonpe** [comment]
> Some thoughts:
> 
> * Any particular reason to not be using the default window size here? The online [examples page](https://bevyengine.org/examples/#stress-tests) just looks nicer when everything is the same aspect ratio.
> * ~What about using `Display::Grid`? Seems like the natural way to lay out a big ol' grid but I guess that changes the characteristics of the test a bit.~ edit: I see you added a grid option while I was reviewing.
> * Basing border size and gaps between buttons on button dimensions effectively erases the borders/gaps with large amounts of buttons.
> 
> Here's a [branch](https://github.com/ickshonpe/bevy/compare/many-buttons-enhancements...rparrett:bevy:9712-grid?expand=1) with `Display::Grid` and static borders / gutters / gaps if you want to experiment with it.

Won't this do the opposite, will the buttons disappear with a lot of buttons?
The margins are slightly different between the grid and flex modes too, it'd be nice to fix that as well.
I guess I don't want to worry too much about the visual aspects and styling in this PR, as it's primarily a benchmark it doesn't matter how it looks in the end. Maybe it would be better to go over them in a second PR.

**Comment by rparrett** [comment]
> Won't this do the opposite, will the buttons disappear with a lot of buttons?

Yeah, that's fair, especially at the particular sizes I chose.

> I guess I don't want to worry too much about the visual aspects and styling in this PR though and as it's primarily a benchmark it doesn't matter how it looks in the end.

I had some vague notion that rendering zero-pixel-width geometry might have different performance characteristics but I doubt that is how it works or that it matters if it is.

> Maybe it would be better to merge this as it is?

Not a blocker for me.

**Comment by rparrett** [comment]
> The margins are slightly different between the grid and flex modes too, it'd be nice to fix that as well.

It might add some additional utility if those modes were producing identical output, but I am not completely convinced that's even possible. Might require a greater flexbox wizard than me.

**Pull request merged**